File: 0057-insert-interval/0057-insert-interval.cpp
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        int n = intervals.size();
        vector<vector<int>> res;
        
        intervals.push_back(newInterval);
        
        sort(intervals.begin(), intervals.end());
        
        // merge overlapping intervals
        for (int i = 0; i < n + 1; i++) { // < n + 1 as we added the newIntervals into our intervals, so its size increases
            if (res.empty() || intervals[i][0] > res.back()[1]) {
                res.push_back(intervals[i]);
            } else {
                res.back()[1] = max(res.back()[1], intervals[i][1]);
            }
        }
        return res;
    }
};

File: 0057-insert-interval/0057-insert-interval.rs
impl Solution {
    pub fn insert(intervals: Vec<Vec<i32>>, new_interval: Vec<i32>) -> Vec<Vec<i32>> {
        let mut intervals = intervals.clone();
        intervals.push(new_interval);
        intervals.sort_by(|a, b| a[0].cmp(&b[0]));
        let mut res: Vec<Vec<i32>> = Vec::new();
        for interval in intervals.iter() {
            if res.is_empty() || res.last().unwrap()[1] < interval[0] {
                res.push(interval.clone());
            } else {
                let last = res.last_mut().unwrap();
                last[1] = std::cmp::max(last[1], interval[1])
            }
        }
        res
    }
}